File: day4/Part1.scala
import java.security.MessageDigest

/**
  * Created by danielchao on 12/10/15.
  */


object Program {
  val input = "bgvyzdsv"

  def md5(s: String) = MessageDigest.getInstance("MD5").digest(s.getBytes("UTF-8")).map("%02x".format(_)).mkString

  def beginsWithFiveZeroes(s: String) = s.startsWith("00000")

  def main() = {
    for (i <- 1 to 1000000) {
      val hashed = md5(input + i.toString)
      if (beginsWithFiveZeroes(hashed)) {
        println(i)
        System.exit(0)
      }
    }
    println("couldnt find answer.")
    System.exit(1)
  }
}

Program.main()

File: 2015/day4/Part1.cpp
#include "md5.h"
#include <iostream>
#include <sstream>

string input("bgvyzdsv");

bool beginsWithFiveZeroes (string str)
{
    return str.substr(0,5) == "00000";
}

int main () {
    for (int i = 0; i < 100000000; i = i + 1)
    {
        string hashed = md5(input + to_string(i));
        if (beginsWithFiveZeroes(hashed))
        {
            cout << to_string(i) << endl;
            return 0;
        }
    }
    cout << "unable to find a string that begins with five zeroes" << endl;
    return 1;
}